providers/implementations/signature/rsa_sig.c

```diff
@@ -105,28 +154,196 @@ typedef struct {
+}
+
+static void wrap_md_freectx(PROV_RSA_CTX *ctx)
+{
+    if (ctx->md.md != NULL)
+        EVP_MD_CTX_free(ctx->mdctx);
+    else if (ctx->md.newctx != NULL)
+        ctx->md.freectx(ctx->mdctx);
+    ctx->mdctx = NULL;
+}
+
+static int wrap_md_init(PROV_RSA_CTX *ctx, const OSSL_PARAM params[])
+{
+    if (ctx->mdctx == NULL)
+        return 0;
+    if (ctx->md.md != NULL)
+        return EVP_DigestInit_ex2(ctx->mdctx, ctx->md.md, params);
+    else if (ctx->md.init != NULL)
+        return ctx->md.init(ctx->mdctx, params);
+    return 0;
+}
+
+static int wrap_md_update(PROV_RSA_CTX *ctx,
+                          const unsigned char *in, size_t inl)
+{
+    if (ctx->mdctx == NULL)
+        return 0;
+    if (ctx->md.md != NULL)
+        return EVP_DigestUpdate(ctx->mdctx, in, inl);
+    else if (ctx->md.update != NULL)
+        return ctx->md.update(ctx->mdctx, in, inl);
+    return 0;
+}
+
+static int wrap_md_final(PROV_RSA_CTX *ctx,
+                         unsigned char *out, size_t *outl, size_t outsz)
+{
+    if (ctx->mdctx == NULL)
+        return 0;
+    if (ctx->md.md != NULL) {
+        unsigned int outlen = (unsigned int)outsz;
+        int ret = EVP_DigestFinal_ex(ctx->mdctx, out, &outlen);
+
+        if (outl != NULL)
+            *outl = (size_t)outlen;
+        return ret;
+    } else if (ctx->md.final != NULL) {
+        return ctx->md.final(ctx->mdctx, out, outl, outsz);
+    }
+    return 0;
+}
+
+static size_t wrap_md_get_size(PROV_RSA_CTX *ctx)
+{
+    if (ctx->md.mdsize > 0)
+        return ctx->md.mdsize;
+
+    if (ctx->md.md != NULL) {
+        int md_size = EVP_MD_get_size(ctx->md.md);
 
-    if (prsactx->md != NULL) {
-        md_size = EVP_MD_get_size(prsactx->md);
         if (md_size <= 0)
             return 0;
         return md_size;
+    } else if (ctx->md.get_params != NULL) {
+        OSSL_PARAM params[2];
+
+        params[0] = OSSL_PARAM_construct_size_t(OSSL_DIGEST_PARAM_SIZE,
+                                                &ctx->md.mdsize);
+        params[1] = OSSL_PARAM_construct_end();
+        if (!ctx->md.get_params(params)) {
+            ctx->md.mdsize = 0;
+            return 0;
+        }
+        return ctx->md.mdsize;
     }
     return 0;
 }
 
+static int wrap_md_up_ref(PROV_RSA_CTX *ctx)
+{
+    if (ctx->md.md != NULL)
+        return EVP_MD_up_ref(ctx->md.md);
+
+    /* Nothing to upref */
+    return 1;
+}
+
+static int wrap_md_is_a(const PROV_RSA_CTX *ctx, const char *mdname, int mdnid)
+{
+    if (ctx->md.md != NULL)
+        return EVP_MD_is_a(ctx->md.md, mdname);
+    return ctx->md.mdnid == mdnid;
+}
+
+static int wrap_md_dup(PROV_RSA_CTX *dstctx, PROV_RSA_CTX *srcctx)
+{
+    if (srcctx->md.md != NULL) {
+        if (!wrap_md_up_ref(srcctx))
+            return 0;
+        dstctx->md.md = srcctx->md.md;
+
+        if (srcctx->mdctx != NULL)
+            if ((dstctx->mdctx = EVP_MD_CTX_new()) == NULL
+                || !EVP_MD_CTX_copy_ex(dstctx->mdctx, srcctx->mdctx)) {
```

Comment:
Can this be collapsed into a call to EVP_MD_CTX_dup()?

Reply:
Oh!  Yes it can.  Apparently, the code I copied here was older...

Done